### PR TITLE
Alternative text for governance

### PR DESCRIPTION
**Pull request by miekg**

### CODE REVIEW

**Comment by miekg** [comment]
Alternative text for governance

@greenpau @fturib

**Comment by fturib** [comment]
Although that is existing behavior - so at least describe how it works so far.
I am not really fan to sign that 'benevolent dictator for life' (I read the wikipedia doc though ..) - it seems a joke for me to write it black on white.

All other parts are good for me.
But I really would like we come back to the 'organization vote'

File: GOVERNANCE.md
Comment (fturib) [inline GOVERNANCE.md:15]: We cannot continue with that as a rule.

File: GOVERNANCE.md
Comment (fturib) [inline GOVERNANCE.md:21]: neither this one.

**Review by fturib — CHANGES_REQUESTED** [review]
That seems a joke ...

File: GOVERNANCE.md
Comment (greenpau) [reply GOVERNANCE.md:21]: @miekg, lgtm :+1

File: GOVERNANCE.md
Comment (miekg) [reply GOVERNANCE.md:15]: Why actually? Not saying I would push for this? But is this a cncf thingy?

Comment (fturib) [reply GOVERNANCE.md:15]: The initial comment was my personal reaction ... 

About the CNCF position, officially it says that it requires only that the governance is stated and clarified. It does not require any particular kind of governance. However, all the projects at Incubator or Graduated level are using votes (considered as an 'open governance').
There were some talk in CNCF about the need to require open-governance (votes for decision) for all projects, or not. As I said above, the official position it to require only that the gobernance is defined.

However, in order to graduate, we need 2/3 of the votes of ToC, and I am thinking that some would refrain to vote if we show the 'dictator'-like governance (that is my personal opinion). Specifically we would be the first project with a non-open governance.

I raised the concern of such governance on last maintainer meeting were Infoblox were represented. I think that Alan/Cricket are reaching to you by email in order to deal the  "Corporate backing" and "Open-governance".
Please, take care to check those email and reply as we need to go ahead quickly on the Governance definition.
Thanks!

**Comment by yongtang** [comment]
I think the issue with coredns is that, coredns is not unique compared with many other open source projects. It is however, might be special inside cncf.

Many other open source projects in cncf have a (or more than one) sponsoring company and the company behind is also the major technical contributor. Companies certainly have more non-technical resources but that typically aligns with technical contributions collectively.

So cncf projects still largely follows merit-based open source model and works well.

For coredns, there is one independent contributor (@miekg) that has more than 50% of the technical contributions day-to-day. There is also a company that may have more non-technical resources, but, simply just could not have enough technical contributions.

Open source software should be merit-based in the area of technical contributions.

For me, it will be a _bigger_ joke if _THE_ major contributor of the project is not involved or does not have a final say on anything in this project.

Note things may change in the future.

**Comment by yongtang** [comment]
A merit-based `project lead` can help incubating, set vision, and reduce noice early on. Things may change when contributions of a project diversify.

But at the current stage, I see more benefits for a `project lead` that drives the direction of the coredns.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Alternative t..." ]
>I think the issue with coredns is that, coredns is not unique compared with many other open source projects. It is however, might be special inside cncf.

this is a very valid statement I think.

>For me, it will be a _bigger_ joke if _THE_ major contributor of the project is not involved or does not have a final say on anything in this project.

We're still discussing, but I value your input. I frankly don't see the BDFL 
would be bad at this stage.

/Miek

--
Miek Gieben

**Comment by greenpau** [comment]
Any thoughts on a _veto_ right? i.e. "Project Lead has the right to exercise a veto against ..."

**Comment by miekg** [comment]
that's a tough one, as I would not expect to be ever put in a situation that would require such a drastic step (but of course a governance doc should preempt that).

If the majority of maintainers think something is a good idea, I don't feel I should cast a veto. 

Should I *technically* be able too? I don't know; what would that accomplish? A fork? Loosing maintainers?

**Comment by fturib** [comment]
> I frankly don't see the BDFL would be bad at this stage.
> Any thoughts on a veto right? i.e. "Project Lead has the right to exercise a veto against ..."

@yongtang proposed this PR with Team Leader : #2220.

Globally that reply I guess those concerns:
- it is a kind a BDFL - but limited in the time. and a vote to renew it. 
- the Team Leader has the last word on any subject -> it is another wording for "veto right".

**Comment by yongtang** [comment]
> Should I technically be able too? I don't know; what would that accomplish? A fork? Loosing maintainers?

It could be many ways. Apache license does not have restriction I believe, so it could be a personal fork, or a fork by a group of maintainers or companies.

In general, open source projects always run with _bona fide_ because there is no restriction. People has the _freedom_ to come and leave so any decision is always technically _non-binding_ to anyone.

In the end, the issue is that the contributions to coredns has not reached to a point of diversification yet (thus more balancing). That is why I think a project lead for coredns makes more sense than the equal vote, because equal vote doesn't work here.

**Comment by yongtang** [comment]
I think in general, I am in supportive of BDFL or some variations. There are several reasons:

First, the lack of diversification of contributions means BDFL is the _de facto_ form of coredns project. As I explained, equal voting does not work here with apache license.

Secondly, for any other forms, who is going to be the enforcer for coredns governance? My concern is that, most of the coredns contributors are independent, and with jobs unrelated to coredns directly. It is hard for independent contributors to stand up in case governance is not followed. They are not going to get paid for coredns, so likely the silent majority. In the end, the governance text does not matter in other forms like equal voting.

Finally, open source project run with _bona fide_, so it comes to who should people trust? I trust @miekg because the majority of the work here is from him so I think he has the most motivation to make coredns project better. One lead also helps clarifying any confusion about governance.

As I explained before, other cncf projects do not have the issue mentioned above. Most other cncf projects have a sponsor company that is also the main contributor. So collectively technical merit is followed and the sponsor company has the motivation to make a project they contribute the most better.

If @miekg some day is hired by the company sponsoring coredns project, then the above problem is solved. I believe that would be a good thing to every one, as it helps avoiding any issue or conflict moving forward. But that certainly is up to individual choice.

There are too may open source projects runs with BDFL and individual contributors are happy to contribute. Any there are so many success names. Here is the long respected list:
https://en.wikipedia.org/wiki/Benevolent_dictator_for_life

As an individual contributor myself, I personally are very happy to contribute in the future for coredns if it is in BDFL.

I do want to say, whatever the governance form is, it needs to be clear up front, and it needs to be followed. Otherwise I don't see the necessity of having a governance text at the first place.

**Comment by yongtang** [comment]
I think in general, I am in supportive of BDFL or some variations. There are several reasons:

First, the lack of diversification of contributions means BDFL is the _de facto_ form of coredns project. As I explained, equal voting does not work here with apache license.

Secondly, for any other forms, who is going to be the enforcer for coredns governance? My concern is that, most of the coredns contributors are independent, and with jobs unrelated to coredns directly. It is hard for independent contributors to stand up in case governance is not followed. They are not going to get paid for coredns, so likely the silent majority. In the end, the governance text does not matter in other forms like equal voting.

Finally, open source project run with _bona fide_, so it comes to who should people trust? I trust @miekg because the majority of the work here is from him so I think he has the most motivation to make coredns project better. One lead also helps clarifying any confusion about governance.

As I explained before, other cncf projects do not have the issue mentioned above. Most other cncf projects have a sponsor company that is also the main contributor. So collectively technical merit is followed and the sponsor company has the motivation to make a project they contribute the most better.

If @miekg some day is hired by the company sponsoring coredns project, then the above problem is solved. I believe that would be a good thing to every one, as it helps avoiding any issue or conflict moving forward. But that certainly is up to individual choice.

There are too may open source projects runs with BDFL and individual contributors are happy to contribute. Any there are so many success names. Here is the long respected list:
https://en.wikipedia.org/wiki/Benevolent_dictator_for_life

As an individual contributor myself, I personally are very happy to contribute in the future for coredns if it is in BDFL.

I do want to say, whatever the governance form is, it needs to be clear up front, and it needs to be followed. Otherwise I don't see the necessity of having a governance text at the first place.

**Comment by yongtang** [comment]
I think in general, I am in supportive of BDFL or some variations. There are several reasons:

First, the lack of diversification of contributions means BDFL is the _de facto_ form of coredns project. As I explained, equal voting does not work here with apache license.

Secondly, for any other forms, who is going to be the enforcer for coredns governance? My concern is that, most of the coredns contributors are independent, and with jobs unrelated to coredns directly. It is hard for independent contributors to stand up in case governance is not followed. They are not going to get paid for coredns, so likely the silent majority. In the end, the governance text does not matter in other forms like equal voting.

Finally, open source project run with _bona fide_, so it comes to who should people trust? I trust @miekg because the majority of the work here is from him so I think he has the most motivation to make coredns project better. One lead also helps clarifying any confusion about governance.

As I explained before, other cncf projects do not have the issue mentioned above. Most other cncf projects have a sponsor company that is also the main contributor. So collectively technical merit is followed and the sponsor company has the motivation to make a project they contribute the most better.

If @miekg some day is hired by the company sponsoring coredns project, then the above problem is solved. I believe that would be a good thing to every one, as it helps avoiding any issue or conflict moving forward. But that certainly is up to individual choice.

There are too may open source projects runs with BDFL and individual contributors are happy to contribute. Any there are so many success names. Here is the long respected list:
https://en.wikipedia.org/wiki/Benevolent_dictator_for_life

As an individual contributor myself, I personally are very happy to contribute in the future for coredns if it is in BDFL.

I do want to say, whatever the governance form is, it needs to be clear up front, and it needs to be followed. Otherwise I don't see the necessity of having a governance text at the first place.

**Review by yongtang — APPROVED** [review]
LGTM. Overall I think BDFL
1) Is clear up front so there is no confusion for any future contributor who wants to help here
2) The governance text with BDFL is more likely to be followed than other ways.

**Comment by greenpau** [comment]
> Finally, open source project run with bona fide, so it comes to who should people trust? I trust @miekg because the majority of the work here is from him so I think he has the most motivation to make coredns project better. One lead also helps clarifying any confusion about governance.

@yongtang , 💯 agree.

**Comment by greenpau** [comment]
> the Team Leader has the last word on any subject -> it is another wording for "veto right".

@fturib , per "has the last word on any subject", I think it is different, because it may mean that the word of a project lead trumps the majority on any subject. The veto right only allows blocking a majority decision. It does not allow advancing an agenda against the majority.

**Comment by yongtang** [comment]
@greenpau I think this is a very good point about `veto`. The definition of `veto` is to `block` an action.

**Review by greenpau — APPROVED** [review]

**Comment by miekg** [comment]
I want to close this PR on move back to #2220 (does that make sense). I think #2020 was now the last

**Pull request closed without merging**
